src/goto-cc/compile.cpp

```diff
@@ -135,56 +136,113 @@ bool compilet::doit()
     warnings_before;
 }
 
+enum class file_typet
+{
+  NO_SUCH_FILE,
+  UNKNOWN,
+  SOURCE_FILE,
+  NORMAL_ARCHIVE,
+  THIN_ARCHIVE,
+  GOTO_BINARY,
+  ELF_OBJECT
+};
+
+static file_typet detect_file_type(const std::string &file_name)
+{
+  // first of all, try to open the file
+  std::ifstream in(file_name);
+  if(!in)
+    return file_typet::NO_SUCH_FILE;
+
+  const std::string::size_type r = file_name.rfind('.', file_name.length() - 1);
```

Comment:
Why not just `file_name.rfind('.')`?

Reply:
Copy&paste from the previous code, I've simplified it now.